Avoid pattern matching in the en/decoders of TxIn/Out

mrBliss opened the pull request:
This avoids roundtripping through `viewCompactTxOut` for no reason, which slows
down writing a snapshot of the ledger state quite a bit.

mrBliss committed:
Avoid pattern matching in the en/decoders of TxIn/Out

This avoids roundtripping through `viewCompactTxOut` for no reason, which slows
down writing a snapshot of the ledger state quite a bit.

mrBliss commented:
I'm assuming there are golden tests for this, correct?

mrBliss commented:
> I'm assuming there are golden tests for this, correct?

The golden tests in consensus passed :relieved:.

mrBliss commented:
From the profiling report of the revalidation of nearly all Shelley blocks:

```
COST CENTRE   SRC                                                %time %alloc
toCBOR        src/Shelley/Spec/Ledger/TxData.hs:(760,3)-(763,20)  10.2   20.5
```
That's the `ToCBOR` instance of `TxOut`. For those wondering why revalidation would ever serialise a `TxOut`: from time to time we write a ledger snapshot (including the entire UTxO) to disk. This happens in a background thread, so I'm not expecting this PR to have a noticeable effect on revalidation speed. Still, it's better to allocate less. Moreover, this will also slightly speed up the decoding of blocks and transactions, as well as past ledger snapshots, speeding up start-up time.

With this PR:
```
COST CENTRE   SRC                                                %time %alloc
toCBOR        src/Shelley/Spec/Ledger/TxData.hs:(760,3)-(763,20)   0.0    0.3
```

mrBliss commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxData.hs:
```diff
@@ -768,8 +768,8 @@ instance
-    (b :: Word64) <- fromCBOR
-    pure $ TxOut addr (Coin $ toInteger b)
+    coin <- fromCBOR
+    pure $ TxOutCompact addr coin
```
Is this bypassing some important validation? If so, and it's crucial, then we should reconsider.

JaredCorduan replied:
decoding as a Word64 is good, that is consistent with how `Coin` is decoded, but we want to still be able to encode `Coin` values that are out of range, ie as an Integer.

JaredCorduan replied:
as @mrBliss said in the other thread, this can't happen since the coin in the output is stored in memory as a Word64.

JaredCorduan commented:
> I'm assuming there are golden tests for this, correct?

yes, there are [golden tests](https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/shelley-spec-ledger-test/test/Test/Shelley/Spec/Ledger/Serialisation/Golden/Encoding.hs#L843), [hedgehog roundtrip tests](https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/shelley-spec-ledger-test/test/Test/Shelley/Spec/Ledger/Serialisation/Tripping/CBOR.hs#L108), and [cddl roundtrip tests](https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/shelley-spec-ledger-test/test/Test/Shelley/Spec/Ledger/Serialisation/CDDL.hs#L81).

mrBliss commented:
> > I'm assuming there are golden tests for this, correct?
> 
> yes, there are [golden tests](https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/shelley-spec-ledger-test/test/Test/Shelley/Spec/Ledger/Serialisation/Golden/Encoding.hs#L843), [hedgehog roundtrip tests](https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/shelley-spec-ledger-test/test/Test/Shelley/Spec/Ledger/Serialisation/Tripping/CBOR.hs#L108), and [cddl roundtrip tests](https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/shelley-spec-ledger-test/test/Test/Shelley/Spec/Ledger/Serialisation/CDDL.hs#L81).

Great. They all passed :slightly_smiling_face:

JaredCorduan commented:
> > > I'm assuming there are golden tests for this, correct?
> > 
> > 
> > yes, there are [golden tests](https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/shelley-spec-ledger-test/test/Test/Shelley/Spec/Ledger/Serialisation/Golden/Encoding.hs#L843), [hedgehog roundtrip tests](https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/shelley-spec-ledger-test/test/Test/Shelley/Spec/Ledger/Serialisation/Tripping/CBOR.hs#L108), and [cddl roundtrip tests](https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/shelley-spec-ledger-test/test/Test/Shelley/Spec/Ledger/Serialisation/CDDL.hs#L81).
> 
> Great. They all passed

I've just realized that the golden tests are too clever, they test coin directly, but "assume the correct thing for types already checked", ie TxOuts, and a roundtrip tests wouldn't notice this kind of change. :( I do think that this change is bypassing the validation on Coin. https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Coin.hs#L47-L54 We still want to serialize the coin values that out out of the intended range (0 to max word64), by encoding them as TkInteger.

mrBliss commented:
> > > > I'm assuming there are golden tests for this, correct?
> > > 
> > > 
> > > yes, there are [golden tests](https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/shelley-spec-ledger-test/test/Test/Shelley/Spec/Ledger/Serialisation/Golden/Encoding.hs#L843), [hedgehog roundtrip tests](https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/shelley-spec-ledger-test/test/Test/Shelley/Spec/Ledger/Serialisation/Tripping/CBOR.hs#L108), and [cddl roundtrip tests](https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/shelley-spec-ledger-test/test/Test/Shelley/Spec/Ledger/Serialisation/CDDL.hs#L81).
> > 
> > 
> > Great. They all passed
> 
> I've just realized that the golden tests are too clever, they test coin directly, but "assume the correct thing for types already checked", ie TxOuts, and a roundtrip tests wouldn't notice this kind of change. :( I do think that this change is bypassing the validation on Coin. https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Coin.hs#L47-L54 We still want to serialize the coin values that out out of the intended range (0 to max word64), by encoding them as TkInteger.

But since `TxOut` stores the coin value already as a `Word64` in memory, `isValidCoinValue` must be true. So I don't think we're missing the validation.

JaredCorduan commented:
> But since `TxOut` stores the coin value already as a `Word64` in memory, `isValidCoinValue` must be true. So I don't think we're missing the validation.

you're totally right.

JaredCorduan reviewed (approved):
thanks @mrBliss ! LGTM

mrBliss commented:
@JaredCorduan Is there any validation in the decoder for `Address` that is crucial?

JaredCorduan commented:
> @JaredCorduan Is there any validation in the decoder for `Address` that is crucial?

Thank you for asking, yes, this is then skipping `getAddr` [here](https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Address.hs#L294)

JaredCorduan reviewed (commented):
I am removing approval, since @mrBliss asked if this could skip any important validation and I do think the address validation is needed

JaredCorduan reviewed (changes requested):
address validation

mrBliss committed:
Validate the address in the FromCBOR instance of TxOut

Slightly more expensive than the previous commit, but still more efficient than
before, as we avoid reserialising the address.

Downside: you won't get a detailed error message like:

    "Address with unknown network Id. (", show addrNetId, ")"

Instead, we use:

    cborError $ DecoderErrorCustom "TxOut" "invalid address"

mrBliss commented:
@JaredCorduan I pushed another commit that adds address validation, albeit in a slightly different way than master, but still much more efficient.

JaredCorduan reviewed (approved):
looks great, thanks again @mrBliss !

mrBliss commented:
bors merge
